File: 498-diagonal-traverse/diagonal-traverse.py

class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        res: List[int] = []
        m = len(mat)
        n = len(mat[0])
        hashmap = {i: 0 for i in range(m)}
        direction = True

        while True:
            iteration_items: List[List[int]] = []
            for key, value in hashmap.items():
                if value == n:
                    continue

                iteration_items.append([key, value])
                hashmap[key] = value + 1
                if value == 0:
                    break

            if direction:
                iteration_items.reverse()
                for pair in iteration_items:
                    res.append(mat[pair[0]][pair[1]])
            else:
                for pair in iteration_items:
                    res.append(mat[pair[0]][pair[1]])

            direction = not direction
            if len(iteration_items) == 0:
                break

        return res

File: 498-diagonal-traverse/diagonal-traverse.cpp
class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
       int m = (int)mat.size();
        if (m == 0) return {};
        int n = (int)mat[0].size();

        vector<int> res;
        res.reserve((size_t)m * n);

        // pos[i] == next column to consume from row i
        vector<int> pos(m, 0);
        bool direction = true; // reverse the collected diagonal when true

        while (true) {
            vector<pair<int,int>> items;
            items.reserve(min(m, n));

            for (int i = 0; i < m; ++i) {
                int v = pos[i];
                if (v == n) continue;           // row exhausted
                items.emplace_back(i, v);
                ++pos[i];
                if (v == 0) break;              // match your Python break
            }

            if (items.empty()) break;

            if (direction) {
                for (int k = (int)items.size() - 1; k >= 0; --k) {
                    auto [r, c] = items[k];
                    res.push_back(mat[r][c]);
                }
            } else {
                for (auto [r, c] : items) {
                    res.push_back(mat[r][c]);
                }
            }
            direction = !direction;
        }
        return res;
    } 
};